src/util/mpscfifo.h

```diff
@@ -1,66 +1,88 @@
 #pragma once
 
 #include <QAtomicInt>
+#include <QMutexLocker>
 
 #include "util/assert.h"
+#include "util/memory.h"
 
-// Lock-free FIFO for multiple producers/writers and a single(!) consumer/reader.
+enum class MpscFifoConcurrency {
+    SingleProducer,
+    MultipleProducers,
+};
+
+// FIFO for multiple producers/writers and a single consumer/reader. Reading
+// is lock-free while concurrent writers are synchronized by a mutex. The
+// mutex can be disabled by explicitly creating a single producer instance.
 template<typename T, int capacity>
 class MpscFifo {
   public:
-    MpscFifo()
-          : m_enqueueSize(0),
-            m_dequeueSize(0),
-            m_headIndex(0),
-            m_tailIndex(0) {
-        static_assert(capacity >= 2, "capacity too low");
+    explicit MpscFifo(
+            MpscFifoConcurrency concurrency = MpscFifoConcurrency::MultipleProducers)
+          : m_writeTokens(0),
+            m_readTokens(capacity),
+            m_writeMutex((concurrency == MpscFifoConcurrency::MultipleProducers) ? new QMutex : nullptr),
+            m_writeIndex(0),
+            m_readIndex(0) {
+        static_assert(capacity >= 1, "capacity too low");
+        static_assert((capacity + 1) > 0, "capacity too high");
     }
 
+    // Writers from multiple threads may enqueue items concurrently.
     bool enqueue(T value) {
-        if (m_enqueueSize.fetchAndAddAcquire(1) >= capacity) {
```

Comment:
Did you consider to pass the value by reference. This easy we have no copy if there is no slot available. It is only relevant if sizeoff T > pointer size though.

Reply:
Conceptually this is a move into the FIFO and passing by value makes this explicit. The case that enqueueing fails should almost never occur when the capacity has been chosen appropriately. That's why I don't want to pass the value by reference here.